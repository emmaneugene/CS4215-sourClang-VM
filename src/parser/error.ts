import * as es from 'estree'

import { ErrorSeverity, ErrorType, SourceError } from '../types'
import { stripIndent } from '../utils/formatters'

export class DisallowedConstructError implements SourceError {
  public type = ErrorType.SYNTAX
  public severity = ErrorSeverity.ERROR
  public nodeType: string

  constructor(public node: es.Node) {
    this.nodeType = this.formatNodeType(this.node.type)
  }

  get location() {
    return this.node.loc!
  }

  public explain() {
    return `${this.nodeType} are not allowed`
  }

  public elaborate() {
    return stripIndent`
      You are trying to use ${this.nodeType}, which is not allowed (yet).
    `
  }

  /**
   * Converts estree node.type into english
   * e.g. ThisExpression -> 'this' expressions
   *      Property -> Properties
   *      EmptyStatement -> Empty Statements
   */
  private formatNodeType(nodeType: string) {
    switch (nodeType) {
      case 'ThisExpression':
        return "'this' expressions"
      case 'Property':
        return 'Properties'
      default: {
        const words = nodeType.split(/(?=[A-Z])/)
        return words.map((word, i) => (i === 0 ? word : word.toLowerCase())).join(' ') + 's'
      }
    }
  }
}

export class FatalSyntaxError implements SourceError {
  public type = ErrorType.SYNTAX
  public severity = ErrorSeverity.ERROR
  public constructor(public location: es.SourceLocation, public message: string) {}

  public explain() {
    return this.message
  }

  public elaborate() {
    return 'There is a syntax error in your program'
  }
}

export class MissingSemicolonError implements SourceError {
  public type = ErrorType.SYNTAX
  public severity = ErrorSeverity.ERROR
  public constructor(public location: es.SourceLocation) {}

  public explain() {
    return 'Missing semicolon at the end of statement'
  }

  public elaborate() {
    return 'Every statement must be terminated by a semicolon.'
  }
}

export class TrailingCommaError implements SourceError {
  public type: ErrorType.SYNTAX
  public severity: ErrorSeverity.WARNING
  public constructor(public location: es.SourceLocation) {}

  public explain() {
    return 'Trailing comma'
  }

  public elaborate() {
    return 'Please remove the trailing comma'
  }
}

export class InvalidConfigError implements SourceError {
  type: ErrorType
  severity: ErrorSeverity
  location: es.SourceLocation

  explain(): string {
    return 'Something went wrong while parsing'
  }

  elaborate(): string {
    return 'Please check the stacktrace'
  }
}

export class InvalidDataType implements SourceError {
  type: ErrorType
  severity: ErrorSeverity
  location: es.SourceLocation

  explain(): string {
    return 'This datatype is not supported'
  }
  elaborate(): string {
    return 'Please check the declared datatype'
  }
}
