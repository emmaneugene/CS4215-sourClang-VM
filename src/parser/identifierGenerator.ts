import { ErrorNode } from 'antlr4ts/tree/ErrorNode'
import { ParseTree } from 'antlr4ts/tree/ParseTree'
import { RuleNode } from 'antlr4ts/tree/RuleNode'
import { TerminalNode } from 'antlr4ts/tree/TerminalNode'
import * as es from 'estree'

import { DeclaratorContext } from '../lang/SourCParser'
import { SourCParserVisitor } from '../lang/SourCParserVisitor'
import { getPointerList } from './utils'

class IdentifierGenerator implements SourCParserVisitor<Omit<es.Identifier, 'datatype'>> {
  visitDeclarator(ctx: DeclaratorContext): Omit<es.Identifier, 'datatype'> {
    if (ctx._array) {
      // array
      throw new Error('Method not implemented')
    }

    if (ctx._funcPointer) {
      // function pointers
      throw new Error('Method not implemented')
    }

    if (ctx._funcProto) {
      // function prototypes
      // TODO: No support for function prototypes yet
      throw new Error('Method not implemented')
    }

    return {
      type: 'Identifier',
      name: ctx.Identifier().text,
      pointerList: getPointerList(ctx.pointer())
    }
  }

  visit(tree: ParseTree): Omit<es.Identifier, 'datatype'> {
    return tree.accept(this)
  }

  visitChildren(node: RuleNode): Omit<es.Identifier, 'datatype'> {
    throw new Error('Method not implemented.')
  }

  visitTerminal(node: TerminalNode): Omit<es.Identifier, 'datatype'> {
    return node.accept(this)
  }

  visitErrorNode(node: ErrorNode): Omit<es.Identifier, 'datatype'> {
    throw new Error('Method not implemented.')
  }
}

export default IdentifierGenerator
