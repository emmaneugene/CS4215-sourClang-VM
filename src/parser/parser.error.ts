import { SourceLocation } from 'estree'

export class FatalSyntaxError extends Error {
  constructor(loc: SourceLocation) {
    const { start, end } = loc
    const message =
      `Error around: ` +
      `Column: ${start.column}-${end.column}.` +
      `Line: ${start.line}-${end.line}`

    super(message)
    this.name = 'FatalSyntaxError'
    this.message = message
  }
}

export class InvalidConfigError extends Error {
  constructor(loc: SourceLocation) {
    const { start, end } = loc
    const message =
      `Error around: ` +
      `Column: ${start.column}-${end.column}.` +
      `Line: ${start.line}-${end.line}`

    super(message)
    this.name = 'InvalidConfigError'
    this.message = message
  }
}
