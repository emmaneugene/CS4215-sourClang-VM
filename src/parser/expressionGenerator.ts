import { ErrorNode } from 'antlr4ts/tree/ErrorNode'
import { ParseTree } from 'antlr4ts/tree/ParseTree'
import { RuleNode } from 'antlr4ts/tree/RuleNode'
import { TerminalNode } from 'antlr4ts/tree/TerminalNode'
import * as es from 'estree'

import {
  AdditiveExpressionContext,
  AssignmentExpressionContext,
  CastExpressionContext,
  ConditionalExpressionContext,
  ConstantExpressionContext,
  EqualityExpressionContext,
  ExpressionContext,
  FunctionCallArgContext,
  InitializerContext,
  InitializerListContext,
  LogicalAndExpressionContext,
  LogicalOrExpressionContext,
  MultiplicativeExpressionContext,
  PostfixExpressionContext,
  PrimaryExpressionContext,
  RelationalExpressionContext,
  UnaryExpressionContext
} from '../lang/SourCParser'
import { SourCParserVisitor } from '../lang/SourCParserVisitor'
import { DataType } from '../typings/datatype'
import { UnaryOperatorContext } from './../lang/SourCParser'
import { InvalidConfigError } from './error'
import { getSizeofOperands, getTypeName } from './utils'

type EsTreeUnaryOperator = '+' | '-' | '!'

function validateUnaryOperator(
  symbol: UnaryOperatorContext | undefined
): EsTreeUnaryOperator | undefined {
  if (!symbol) {
    return undefined
  }

  if (symbol.Plus()) {
    return '+'
  } else if (symbol.Minus()) {
    return '-'
  } else if (symbol.Not()) {
    return '!'
  }

  return undefined
}

class ExpressionGenerator implements SourCParserVisitor<es.Expression> {
  visitAssignmentExpression(ctx: AssignmentExpressionContext): es.Expression {
    if (ctx.conditionalExpression()) {
      return ctx.conditionalExpression()!.accept(this)
    }

    if (ctx.unaryExpression() && ctx.Assign() && ctx.assignmentExpression()) {
      return {
        type: 'FlexiAssignmentExpression',
        operator: '=', // TODO: Map the correct operator over
        left: ctx.unaryExpression()!.accept(this),
        right: ctx.assignmentExpression()!.accept(this)
      }
    }

    throw new InvalidConfigError()
  }

  visitConditionalExpression(ctx: ConditionalExpressionContext): es.Expression {
    if (ctx.Question() && ctx.Colon()) {
      return {
        type: 'ConditionalExpression',
        test: ctx.logicalOrExpression().accept(this),
        consequent: ctx._cons.accept(this),
        alternate: ctx._alt.accept(this)
      }
    }

    return ctx.logicalOrExpression().accept(this)
  }

  visitLogicalOrExpression(ctx: LogicalOrExpressionContext): es.Expression {
    if (ctx.OrOr() && ctx.logicalOrExpression()) {
      return {
        type: 'LogicalExpression',
        operator: '||',
        left: ctx.logicalOrExpression()!.accept(this),
        right: ctx.logicalAndExpression().accept(this)
      }
    }

    return ctx.logicalAndExpression().accept(this)
  }

  visitLogicalAndExpression(ctx: LogicalAndExpressionContext): es.Expression {
    if (ctx.AndAnd() && ctx.logicalAndExpression()) {
      return {
        type: 'LogicalExpression',
        operator: '&&',
        left: ctx.logicalAndExpression()!.accept(this),
        right: ctx.equalityExpression().accept(this)
      }
    }

    return ctx.equalityExpression().accept(this)
  }

  visitEqualityExpression(ctx: EqualityExpressionContext): es.Expression {
    if (ctx.Equal() && ctx.equalityExpression()) {
      return {
        type: 'BinaryExpression',
        operator: '==',
        left: ctx.equalityExpression()!.accept(this),
        right: ctx.relationalExpression().accept(this)
      }
    }

    if (ctx.NotEqual() && ctx.equalityExpression()) {
      return {
        type: 'BinaryExpression',
        operator: '!=',
        left: ctx.equalityExpression()!.accept(this),
        right: ctx.relationalExpression().accept(this)
      }
    }

    return ctx.relationalExpression().accept(this)
  }

  visitRelationalExpression(ctx: RelationalExpressionContext): es.Expression {
    if (ctx.Less() && ctx.relationalExpression()) {
      return {
        type: 'BinaryExpression',
        operator: '<',
        left: ctx.relationalExpression()!.accept(this),
        right: ctx.additiveExpression().accept(this)
      }
    }

    if (ctx.Greater() && ctx.relationalExpression()) {
      return {
        type: 'BinaryExpression',
        operator: '>',
        left: ctx.relationalExpression()!.accept(this),
        right: ctx.additiveExpression().accept(this)
      }
    }

    if (ctx.LessEqual() && ctx.relationalExpression()) {
      return {
        type: 'BinaryExpression',
        operator: '<=',
        left: ctx.relationalExpression()!.accept(this),
        right: ctx.additiveExpression().accept(this)
      }
    }

    if (ctx.GreaterEqual() && ctx.relationalExpression()) {
      return {
        type: 'BinaryExpression',
        operator: '>=',
        left: ctx.relationalExpression()!.accept(this),
        right: ctx.additiveExpression().accept(this)
      }
    }

    return ctx.additiveExpression().accept(this)
  }

  visitAdditiveExpression(ctx: AdditiveExpressionContext): es.Expression {
    if (ctx.Plus() && ctx.additiveExpression()) {
      return {
        type: 'BinaryExpression',
        operator: '+',
        left: ctx.additiveExpression()!.accept(this),
        right: ctx.multiplicativeExpression().accept(this)
      }
    }

    if (ctx.Minus() && ctx.additiveExpression()) {
      return {
        type: 'BinaryExpression',
        operator: '-',
        left: ctx.additiveExpression()!.accept(this),
        right: ctx.multiplicativeExpression().accept(this)
      }
    }

    return ctx.multiplicativeExpression().accept(this)
  }

  visitMultiplicativeExpression(ctx: MultiplicativeExpressionContext): es.Expression {
    if (ctx.Star() && ctx.multiplicativeExpression()) {
      return {
        type: 'BinaryExpression',
        operator: '*',
        left: ctx.multiplicativeExpression()!.accept(this),
        right: ctx.castExpression().accept(this)
      }
    }

    if (ctx.Div() && ctx.multiplicativeExpression()) {
      return {
        type: 'BinaryExpression',
        operator: '/',
        left: ctx.multiplicativeExpression()!.accept(this),
        right: ctx.castExpression().accept(this)
      }
    }

    if (ctx.Mod() && ctx.multiplicativeExpression()) {
      return {
        type: 'BinaryExpression',
        operator: '%',
        left: ctx.multiplicativeExpression()!.accept(this),
        right: ctx.castExpression().accept(this)
      }
    }

    return ctx.castExpression().accept(this)
  }

  visitCastExpression(ctx: CastExpressionContext): es.Expression {
    if (ctx.LeftParen() && ctx.typeName() && ctx.RightParen() && ctx.castExpression()) {
      return {
        type: 'CastExpression',
        targetType: getTypeName(ctx.typeName()!),
        expression: ctx.castExpression()!.accept(this)
      }
    }

    if (ctx.unaryExpression()) {
      return ctx.unaryExpression()!.accept(this)
    }

    throw new InvalidConfigError()
  }

  visitUnaryExpression(ctx: UnaryExpressionContext): es.Expression {
    if (ctx.postfixExpression()) {
      return ctx.postfixExpression()!.accept(this)
    }

    if (ctx.PlusPlus() && ctx.unaryOperator() && ctx.unaryExpression()) {
      return {
        type: 'UpdateExpression',
        operator: '++',
        argument: ctx.unaryExpression()!.accept(this),
        prefix: true // we set this to false in postfixExpression
      }
    }

    if (ctx.MinusMinus() && ctx.unaryOperator() && ctx.unaryExpression()) {
      return {
        type: 'UpdateExpression',
        operator: '--',
        argument: ctx.unaryExpression()!.accept(this),
        prefix: true // we set this to false in visitPostfixExpression
      }
    }

    if (ctx.Sizeof() && ctx.sizeofOperands()) {
      return {
        type: 'SizeofExpression',
        operand: getSizeofOperands(ctx.sizeofOperands()!)
      }
    }

    if (ctx.unaryOperator() && ctx.castExpression()) {
      const op = ctx.unaryOperator()!

      if (op.Star()) {
        return {
          type: 'ValueofExpression',
          expression: ctx.castExpression()!.accept(this)
        }
      } else if (op.And()) {
        return {
          type: 'AddressofExpression',
          expression: ctx.castExpression()!.accept(this)
        }
      } else if (validateUnaryOperator(op)) {
        return {
          type: 'UnaryExpression',
          operator: validateUnaryOperator(op)!,
          prefix: true,
          argument: ctx.castExpression()!.accept(this)
        }
      } else {
        throw new Error('Method not implemented yet')
      }
    }

    throw new InvalidConfigError()
  }

  visitPostfixExpression(ctx: PostfixExpressionContext): es.Expression {
    if (ctx.primaryExpression()) {
      return ctx.primaryExpression()!.accept(this)
    }

    if (!ctx.postfixExpression()) {
      // Guard clause for the rest of the if statements
      throw new InvalidConfigError()
    }

    if (ctx.LeftBracket() && ctx.expression() && ctx.RightBracket()) {
      // Array access
      return {
        type: 'MemberExpression',
        object: ctx.postfixExpression()!.accept(this),
        property: ctx.expression()!.accept(this),
        computed: true, // If this is true, it is array access
        optional: false // Not sure what this means in the JS context
      }
    }

    if (ctx.LeftParen() && ctx.functionCallArg() && ctx.RightParen()) {
      // Function call
      return {
        type: 'CallExpression',
        callee: ctx.postfixExpression()!.accept(this),
        arguments: this.visitFunctionCallArg(ctx.functionCallArg()!).expressions,
        optional: false // Not sure what this means in the JS context
      }
    }

    if (ctx.Dot() && ctx.Identifier()) {
      // Direct struct access
      return {
        type: 'MemberExpression',
        object: ctx.postfixExpression()!.accept(this),
        property: {
          type: 'Identifier',
          name: ctx.Identifier()!.text,
          datatype: DataType.UNKNOWN // to be determined from the struct def
        },
        computed: false, // If this is false, it is member access,
        optional: false
      }
    }

    if (ctx.Arrow() && ctx.Identifier()) {
      // TODO: Struct access from memory address not supported yet
      throw new InvalidConfigError()
    }

    if (ctx.PlusPlus() && ctx.postfixExpression()) {
      return {
        type: 'UpdateExpression',
        operator: '++',
        argument: ctx.postfixExpression()!.accept(this),
        prefix: false
      }
    }

    if (ctx.MinusMinus() && ctx.postfixExpression()) {
      return {
        type: 'UpdateExpression',
        operator: '--',
        argument: ctx.postfixExpression()!.accept(this),
        prefix: false
      }
    }

    throw new InvalidConfigError()
  }

  visitFunctionCallArg(ctx: FunctionCallArgContext): es.SequenceExpression {
    if (ctx.Comma() && ctx.functionCallArg()) {
      return {
        type: 'SequenceExpression',
        expressions: [
          ...this.visitFunctionCallArg(ctx.functionCallArg()!).expressions,
          ctx.assignmentExpression().accept(this)
        ]
      }
    }

    return {
      type: 'SequenceExpression',
      expressions: [ctx.assignmentExpression().accept(this)]
    }
  }

  visitPrimaryExpression(ctx: PrimaryExpressionContext): es.Expression {
    if (ctx.Identifier()) {
      return {
        type: 'Identifier',
        name: ctx.Identifier()!.text,
        datatype: DataType.UNKNOWN
      }
    }

    if (ctx.Constant()) {
      return {
        type: 'Literal',
        value: parseInt(ctx.Constant()!.text),
        raw: ctx.Constant()!.text
      }
    }

    if (ctx.StringLiteral()) {
      return {
        type: 'Literal',
        value: ctx.StringLiteral()!.text,
        raw: ctx.StringLiteral()!.text
      }
    }

    if (ctx.LeftParen() && ctx.expression() && ctx.RightParen()) {
      return ctx.expression()!.accept(this)
    }

    throw new InvalidConfigError()
  }

  /**
   * Most of the action for the ExpressionGenerator is here.
   * It is up to the context/parent to determine if this makes semantic sense.
   */
  visitExpression(ctx: ExpressionContext): es.Expression {
    return ctx.assignmentExpression().accept(this)
  }

  visitConstantExpression(ctx: ConstantExpressionContext): es.Expression {
    return ctx.conditionalExpression().accept(this)
  }

  visitInitializer(ctx: InitializerContext): es.Expression {
    if (ctx.LeftBrace() && ctx.initializerList() && ctx.RightBrace()) {
      // See `visitInitializerList`. This returns `es.SequenceExpression`
      return ctx.initializerList()!.accept(this)
    }

    if (ctx.assignmentExpression()) {
      return ctx.assignmentExpression()!.accept(this)
    }

    throw new InvalidConfigError()
  }

  /* The main point of this method is to flatten the recursive definition */
  visitInitializerList(ctx: InitializerListContext): es.Expression {
    // base case
    if (!ctx.Comma()) {
      return {
        type: 'SequenceExpression',
        expressions: [ctx.initializer().accept(this)]
      }
    }

    if (ctx.initializerList()) {
      const remaining = ctx.initializerList()!.accept(this) as es.SequenceExpression
      return {
        type: 'SequenceExpression',
        expressions: [...remaining.expressions, ctx.initializer().accept(this)]
      }
    }

    throw new InvalidConfigError()
  }

  visit(tree: ParseTree): es.Expression {
    return tree.accept(this)
  }

  visitChildren(node: RuleNode): es.Expression {
    throw new Error('Method not implemented.')
  }

  visitTerminal(node: TerminalNode): es.Expression {
    return node.accept(this)
  }

  visitErrorNode(node: ErrorNode): es.Expression {
    throw new Error('Method not implemented.')
  }
}

export default ExpressionGenerator
