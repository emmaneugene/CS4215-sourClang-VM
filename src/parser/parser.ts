import { CharStreams, CommonTokenStream } from 'antlr4ts'
import * as es from 'estree'

import { SourCLexer } from '../lang/SourCLexer'
import { Context, ErrorSeverity } from '../types'
import { SourCParser, TranslationUnitContext } from './../lang/SourCParser'
import { FatalSyntaxError } from './error'
import StatementGenerator from './statementGenerator'

function convertStatement(translationUnit: TranslationUnitContext): es.Statement {
  const generator = new StatementGenerator()
  return translationUnit.accept(generator)
}

function convertSource(translationUnit: TranslationUnitContext): es.Program {
  return {
    type: 'Program',
    sourceType: 'script',
    body: [convertStatement(translationUnit)]
  }
}

export function parse(source: string, context: Context) {
  let program: es.Program | undefined

  if (context.variant === 'calc') {
    const inputStream = CharStreams.fromString(source)
    const lexer = new SourCLexer(inputStream)
    const tokenStream = new CommonTokenStream(lexer)
    const parser = new SourCParser(tokenStream)
    parser.buildParseTree = true
    try {
      const tree = parser.translationUnit()
      program = convertSource(tree)
    } catch (error) {
      if (error instanceof FatalSyntaxError) {
        context.errors.push(error)
      } else {
        throw error
      }
    }
    const hasErrors = context.errors.find(m => m.severity === ErrorSeverity.ERROR)
    if (program && !hasErrors) {
      console.log(JSON.stringify(program, null, 2))
      return program
    } else {
      return undefined
    }
  } else {
    return undefined
  }
}

/*
import { CalcLexer } from '../lang/CalcLexer'
import {
  AdditionContext,
  CalcParser,
  DivisionContext,
  ExpressionContext,
  MultiplicationContext,
  NumberContext,
  ParenthesesContext,
  PowerContext,
  StartContext,
  SubtractionContext
} from '../lang/CalcParser'
import { CalcVisitor } from '../lang/CalcVisitor'

class ExpressionGenerator implements CalcVisitor<es.Expression> {
  visitNumber(ctx: NumberContext): es.Expression {
    return {
      type: 'Literal',
      value: parseInt(ctx.text),
      raw: ctx.text,
      loc: contextToLocation(ctx)
    }
  }
  visitParentheses(ctx: ParenthesesContext): es.Expression {
    return this.visit(ctx.expression())
  }
  visitPower(ctx: PowerContext): es.Expression {
    return {
      type: 'BinaryExpression',
      operator: '^',
      left: this.visit(ctx._left),
      right: this.visit(ctx._right),
      loc: contextToLocation(ctx)
    }
  }

  visitMultiplication(ctx: MultiplicationContext): es.Expression {
    return {
      type: 'BinaryExpression',
      operator: '*',
      left: this.visit(ctx._left),
      right: this.visit(ctx._right),
      loc: contextToLocation(ctx)
    }
  }
  visitDivision(ctx: DivisionContext): es.Expression {
    return {
      type: 'BinaryExpression',
      operator: '/',
      left: this.visit(ctx._left),
      right: this.visit(ctx._right),
      loc: contextToLocation(ctx)
    }
  }
  visitAddition(ctx: AdditionContext): es.Expression {
    return {
      type: 'BinaryExpression',
      operator: '+',
      left: this.visit(ctx._left),
      right: this.visit(ctx._right),
      loc: contextToLocation(ctx)
    }
  }

  visitSubtraction(ctx: SubtractionContext): es.Expression {
    return {
      type: 'BinaryExpression',
      operator: '-',
      left: this.visit(ctx._left),
      right: this.visit(ctx._right),
      loc: contextToLocation(ctx)
    }
  }

  visitExpression?: ((ctx: ExpressionContext) => es.Expression) | undefined
  visitStart?: ((ctx: StartContext) => es.Expression) | undefined

  visit(tree: ParseTree): es.Expression {
    return tree.accept(this)
  }
  visitChildren(node: RuleNode): es.Expression {
    const expressions: es.Expression[] = []
    for (let i = 0; i < node.childCount; i++) {
      expressions.push(node.getChild(i).accept(this))
    }
    return {
      type: 'SequenceExpression',
      expressions
    }
  }
  visitTerminal(node: TerminalNode): es.Expression {
    return node.accept(this)
  }

  visitErrorNode(node: ErrorNode): es.Expression {
    throw new FatalSyntaxError(
      {
        start: {
          line: node.symbol.line,
          column: node.symbol.charPositionInLine
        },
        end: {
          line: node.symbol.line,
          column: node.symbol.charPositionInLine + 1
        }
      },
      `invalid syntax ${node.text}`
    )
  }
}

function convertExpression(expression: ExpressionContext): es.Expression {
  const generator = new ExpressionGenerator()
  return expression.accept(generator)
}
*/
