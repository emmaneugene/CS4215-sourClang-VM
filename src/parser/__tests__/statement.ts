import { BlockStatement, FunctionDeclaration, Program } from 'estree'

import createContext from '../../createContext'
import { Context, Variant } from '../../types'
import { parse } from '../parser'

const f = (when: string, should: string) => `When_\'${when}\'_Should_${should}`
const prependFixture = (stmt: string) => `int main() { ${stmt} }`
const getStmt = (result: Program | undefined) => {
  return ((result?.body[0] as BlockStatement)?.body[0] as FunctionDeclaration).body.body[0]
}

let context: Context = createContext(Variant.DEFAULT, undefined, undefined)

beforeAll(() => {
  console.log = () => {}
})

beforeEach(() => {
  context = createContext(Variant.DEFAULT, undefined, undefined)
})

const IF_STMT: string[] = [
  'if (x == 1) { x = 10; } ',
  'if (x == -1) { x = 10; } else { x = x - 1; } '
]

describe('If', () => {
  for (const source of IF_STMT) {
    it(f(source, 'ParseSuccessfully'), () => {
      const result = parse(prependFixture(source), context)
      expect(getStmt(result)).toMatchSnapshot()
    })
  }
})

const WHILE_STMT: string[] = [
  'while (x < 5) { x++; }',
  'while (1) { }',
  'while (10 + 5 < x) { x--; }',
  'while (1) { break; }',
  'while (1) { continue; }',
  'while (1) { return -x + 10; }'
]

describe('While', () => {
  for (const source of WHILE_STMT) {
    it(f(source, 'ParseSuccessfully'), () => {
      const result = parse(prependFixture(source), context)
      expect(getStmt(result)).toMatchSnapshot()
    })
  }
})

const FOR_STMT: string[] = ['for (x = 1; x < 10; x++) { y++; }', 'for(;;) { continue; }']

describe('For', () => {
  for (const source of FOR_STMT) {
    it(f(source, 'ParseSuccessfully'), () => {
      const result = parse(prependFixture(source), context)
      expect(getStmt(result)).toMatchSnapshot()
    })
  }
})

const FUNCTION_DEF: string[] = ['int foo() { return 10;} ']

describe('FunctionDefinition', () => {
  for (const source of FUNCTION_DEF) {
    it(f(source, 'ParseSuccessfully'), () => {
      const result = parse(source, context)
      expect(result).toMatchSnapshot()
    })
  }
})
