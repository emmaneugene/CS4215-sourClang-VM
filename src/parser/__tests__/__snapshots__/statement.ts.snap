// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`For When_'for (x = 1; x < 10; x++) { y++; }'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Object {
    "body": Array [
      Object {
        "expression": Object {
          "argument": Object {
            "datatype": "unknown",
            "name": "y",
            "type": "Identifier",
          },
          "operator": "++",
          "prefix": false,
          "type": "UpdateExpression",
        },
        "type": "ExpressionStatement",
      },
    ],
    "type": "BlockStatement",
  },
  "init": Object {
    "left": Object {
      "datatype": "unknown",
      "name": "x",
      "type": "Identifier",
    },
    "operator": "=",
    "right": Object {
      "raw": "1",
      "type": "Literal",
      "value": 1,
    },
    "type": "FlexiAssignmentExpression",
  },
  "test": Object {
    "left": Object {
      "datatype": "unknown",
      "name": "x",
      "type": "Identifier",
    },
    "operator": "<",
    "right": Object {
      "raw": "10",
      "type": "Literal",
      "value": 10,
    },
    "type": "BinaryExpression",
  },
  "type": "ForStatement",
  "update": Object {
    "argument": Object {
      "datatype": "unknown",
      "name": "x",
      "type": "Identifier",
    },
    "operator": "++",
    "prefix": false,
    "type": "UpdateExpression",
  },
}
`;

exports[`For When_'for(;;) { continue; }'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Object {
    "body": Array [
      Object {
        "type": "ContinueStatement",
      },
    ],
    "type": "BlockStatement",
  },
  "init": undefined,
  "test": undefined,
  "type": "ForStatement",
  "update": undefined,
}
`;

exports[`FunctionDefinition When_'int foo() { return 10;} '_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "body": Object {
            "body": Array [
              Object {
                "argument": Object {
                  "raw": "10",
                  "type": "Literal",
                  "value": 10,
                },
                "type": "ReturnStatement",
              },
            ],
            "type": "BlockStatement",
          },
          "id": Object {
            "datatype": "int",
            "name": "foo",
            "type": "Identifier",
          },
          "params": Array [],
          "type": "FunctionDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`If When_'if (x == -1) { x = 10; } else { x = x - 1; } '_Should_ParseSuccessfully 1`] = `
Object {
  "alternate": Object {
    "body": Array [
      Object {
        "expression": Object {
          "left": Object {
            "datatype": "unknown",
            "name": "x",
            "type": "Identifier",
          },
          "operator": "=",
          "right": Object {
            "left": Object {
              "datatype": "unknown",
              "name": "x",
              "type": "Identifier",
            },
            "operator": "-",
            "right": Object {
              "raw": "1",
              "type": "Literal",
              "value": 1,
            },
            "type": "BinaryExpression",
          },
          "type": "FlexiAssignmentExpression",
        },
        "type": "ExpressionStatement",
      },
    ],
    "type": "BlockStatement",
  },
  "consequent": Object {
    "body": Array [
      Object {
        "expression": Object {
          "left": Object {
            "datatype": "unknown",
            "name": "x",
            "type": "Identifier",
          },
          "operator": "=",
          "right": Object {
            "raw": "10",
            "type": "Literal",
            "value": 10,
          },
          "type": "FlexiAssignmentExpression",
        },
        "type": "ExpressionStatement",
      },
    ],
    "type": "BlockStatement",
  },
  "test": Object {
    "left": Object {
      "datatype": "unknown",
      "name": "x",
      "type": "Identifier",
    },
    "operator": "==",
    "right": Object {
      "argument": Object {
        "raw": "1",
        "type": "Literal",
        "value": 1,
      },
      "operator": "-",
      "prefix": true,
      "type": "UnaryExpression",
    },
    "type": "BinaryExpression",
  },
  "type": "IfStatement",
}
`;

exports[`If When_'if (x == 1) { x = 10; } '_Should_ParseSuccessfully 1`] = `
Object {
  "alternate": undefined,
  "consequent": Object {
    "body": Array [
      Object {
        "expression": Object {
          "left": Object {
            "datatype": "unknown",
            "name": "x",
            "type": "Identifier",
          },
          "operator": "=",
          "right": Object {
            "raw": "10",
            "type": "Literal",
            "value": 10,
          },
          "type": "FlexiAssignmentExpression",
        },
        "type": "ExpressionStatement",
      },
    ],
    "type": "BlockStatement",
  },
  "test": Object {
    "left": Object {
      "datatype": "unknown",
      "name": "x",
      "type": "Identifier",
    },
    "operator": "==",
    "right": Object {
      "raw": "1",
      "type": "Literal",
      "value": 1,
    },
    "type": "BinaryExpression",
  },
  "type": "IfStatement",
}
`;

exports[`While When_'while (1) { }'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Object {
    "body": Array [],
    "type": "BlockStatement",
  },
  "test": Object {
    "raw": "1",
    "type": "Literal",
    "value": 1,
  },
  "type": "WhileStatement",
}
`;

exports[`While When_'while (1) { break; }'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Object {
    "body": Array [
      Object {
        "type": "BreakStatement",
      },
    ],
    "type": "BlockStatement",
  },
  "test": Object {
    "raw": "1",
    "type": "Literal",
    "value": 1,
  },
  "type": "WhileStatement",
}
`;

exports[`While When_'while (1) { continue; }'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Object {
    "body": Array [
      Object {
        "type": "ContinueStatement",
      },
    ],
    "type": "BlockStatement",
  },
  "test": Object {
    "raw": "1",
    "type": "Literal",
    "value": 1,
  },
  "type": "WhileStatement",
}
`;

exports[`While When_'while (1) { return -x + 10; }'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Object {
    "body": Array [
      Object {
        "argument": Object {
          "left": Object {
            "argument": Object {
              "datatype": "unknown",
              "name": "x",
              "type": "Identifier",
            },
            "operator": "-",
            "prefix": true,
            "type": "UnaryExpression",
          },
          "operator": "+",
          "right": Object {
            "raw": "10",
            "type": "Literal",
            "value": 10,
          },
          "type": "BinaryExpression",
        },
        "type": "ReturnStatement",
      },
    ],
    "type": "BlockStatement",
  },
  "test": Object {
    "raw": "1",
    "type": "Literal",
    "value": 1,
  },
  "type": "WhileStatement",
}
`;

exports[`While When_'while (10 + 5 < x) { x--; }'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Object {
    "body": Array [
      Object {
        "expression": Object {
          "argument": Object {
            "datatype": "unknown",
            "name": "x",
            "type": "Identifier",
          },
          "operator": "--",
          "prefix": false,
          "type": "UpdateExpression",
        },
        "type": "ExpressionStatement",
      },
    ],
    "type": "BlockStatement",
  },
  "test": Object {
    "left": Object {
      "left": Object {
        "raw": "10",
        "type": "Literal",
        "value": 10,
      },
      "operator": "+",
      "right": Object {
        "raw": "5",
        "type": "Literal",
        "value": 5,
      },
      "type": "BinaryExpression",
    },
    "operator": "<",
    "right": Object {
      "datatype": "unknown",
      "name": "x",
      "type": "Identifier",
    },
    "type": "BinaryExpression",
  },
  "type": "WhileStatement",
}
`;

exports[`While When_'while (x < 5) { x++; }'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Object {
    "body": Array [
      Object {
        "expression": Object {
          "argument": Object {
            "datatype": "unknown",
            "name": "x",
            "type": "Identifier",
          },
          "operator": "++",
          "prefix": false,
          "type": "UpdateExpression",
        },
        "type": "ExpressionStatement",
      },
    ],
    "type": "BlockStatement",
  },
  "test": Object {
    "left": Object {
      "datatype": "unknown",
      "name": "x",
      "type": "Identifier",
    },
    "operator": "<",
    "right": Object {
      "raw": "5",
      "type": "Literal",
      "value": 5,
    },
    "type": "BinaryExpression",
  },
  "type": "WhileStatement",
}
`;
